feat(metrics): Add validator metrics for cpu model and system memory (#4666)

nmrshll opened the pull request:
# Description of change

Introduces enhancements to the validator metrics by adding support for getting validator hardware specs, like CPU model/cores, system memory, disk space. This data is sent to iota-proxy (for us to monitor the state of the network through telemetry).

Also fixes private network setup scripts/configs, which didn't seem to work with the exact commands given in the different READMEs (I encountered the same problems as described in #5204, had to fix them to test this PR with docker)

Added example grafana dashboard items:

<img width="962" alt="image" src="https://github.com/user-attachments/assets/d7013e53-da07-41df-a1e9-7045ed5fc63e" />

## Links to any relevant issues

Fixes #4666.

## Platform support:

- Linux / MacOS, arm64/x86
- ⚠️ Unfortunately, for docker on MacOS (and same for Orbstack and podman-machine), the stats gathered cannot be reliable (at least without complicating this PR and the docker setup): cpu model usually missing, disk space shows any value provided by the docker virtual machine
- on docker on Linux (which is what most people running our default node setup will likely run), the stats are gathered correctly

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

The changes have been tested in a local environment to ensure that the new metrics are being collected correctly and that the private network setup works seamlessly with Grafana. Further testing will be conducted based on feedback received.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

nmrshll committed:
WIP 4666

nmrshll committed:
dashboard query

nmrshll committed:
wip

nmrshll committed:
cleanup; add license

nmrshll committed:
formatting

nmrshll committed:
fix tests

nmrshll committed:
fmt

nmrshll committed:
format

nmrshll committed:
Merge branch 'develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

alexsporn commented on crates/iota-proxy/src/lib.rs:
```diff
@@ -203,7 +203,7 @@ mod tests {
 
     /// this is a long test to ensure we are timing out clients that are slow
-    #[tokio::test]
+    #[tokio::test(flavor = "current_thread", start_paused = true)] // start_paused means time is simulated, progresses without waiting
```
why is this needed?

nmrshll replied:
this was to try and skip the wait for the timeout test. It doesn't seem to work as simple as that so I've reverted for now.

alexsporn commented on crates/iota-proxy/src/main.rs:
```diff
@@ -100,6 +100,7 @@ async fn main() -> Result<()> {
             "unavailable",
         ))
         .unwrap();
+
```
revert?

nmrshll replied:
reverted now

alexsporn commented on consensus/config/src/committee.rs:
```diff
@@ -45,10 +45,10 @@ impl Committee {
         assert_ne!(total_stake, 0, "Total stake cannot be zero!");
         let quorum_threshold = 2 * total_stake / 3 + 1;
-        let validity_threshold = (total_stake + 2) / 3;
+        let validity_threshold = total_stake.div_ceil(3);
```
where do these changes come from?

nmrshll replied:
Locally clippy surfaces many more lints for me. Seems like maybe it doesn't pick up the `.clippy.toml` config ? I haven't figured out why yet, I'll look into it.
I've reverted this for now.

alexsporn commented on crates/iota-metrics/Cargo.toml:
```diff
@@ -23,10 +23,15 @@ parking_lot.workspace = true
 prometheus-closure-metric.workspace = true
+
+[dev-dependencies]
+snap.workspace = true
```
this seems unused

nmrshll replied:
removed it now

nmrshll committed:
revert tokio paused

nmrshll committed:
revert fmt

nmrshll committed:
added cpu arch, is_docker, os version

nmrshll committed:
fmt

nmrshll committed:
revert dependencies

nmrshll committed:
revert clippy fix

nmrshll committed:
Merge branch 'develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

nmrshll committed:
Merge branch 'develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

DaughterOfMars commented:
Can we fix the title? Should be conventional commit format

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory_backup_2503121125

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

DaughterOfMars commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,283 @@
+            .unwrap();
+        assert!(core_count > 0 && core_count < 513);
+
+        let mem_total_human = find_metric_label("memory_specs", "memory_total_human");
+        assert!(
+            mem_total_human.ends_with(" GB"),
+            "memory is usually in the GB range at the moment"
+        );
+
+        let disk_total_human = find_metric_label("disk_specs", "disk_total_space_human");
+        assert!(
+            disk_total_human.ends_with(" GB") || disk_total_human.ends_with(" TB"),
+            "disk size is usually in GB or TB at the moment"
+        );
```
I don't know about these tests, they seem like they could start failing randomly. Maybe that's very unlikely but I'd still rather have something future-proof

nmrshll replied:
no problem, I'll remove them and just check the value is non-zero

DaughterOfMars commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2024 IOTA Stiftung
```
```suggestion
// Copyright (c) 2025 IOTA Stiftung
```

muXxer commented on consensus/config/src/committee.rs:
```diff
@@ -45,7 +45,7 @@ impl Committee {
         );
 
-        let total_stake = authorities.iter().map(|a| a.stake).sum();
+        let total_stake: u64 = authorities.iter().map(|a| a.stake).sum();
```
why that change? I think we shouldn't touch the consensus files :)

nmrshll replied:
leftover from a change suggested by clippy (which wasn't configured the same on my machine as in CI)
reverted it now 👍

nmrshll committed:
fix: PR feedback

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

muXxer commented on crates/iota-proxy/Cargo.toml:
```diff
@@ -36,7 +36,7 @@ serde_json.workspace = true
 serde_with.workspace = true
 serde_yaml.workspace = true
 snap.workspace = true
-tokio = { workspace = true, features = ["full"] }
```
why that change? don't we need the full features here?

nmrshll replied:
reverted

muXxer commented on dev-tools/iota-private-network/docker-compose.yaml:
```diff
@@ -1,7 +1,7 @@
 
 x-common-validator: &common-validator
-  image: iota-node
+  image: iotaledger/iota-node
```
IMO these changes should not be part of this PR. We should open another PR for that.

The reasoning to not use the `iotaledger/` images was, that you have to use the "self-built" ones which can be built using the scripts in the repo. We could change everything to `iotaledger/` if we want, but then we should also change that in the build scripts, so that you either pull the latest image from docker hub, or just overwrite it with your local build. But then we should update that everywhere in the repo.

nmrshll replied:
If I understand correctly that's what was broken in #5204 . The docker build tags images with `iotaledger/`
If I use image tags without `iotaledger/` I need to change the build scripts as well

Anyway, I'll move this to its own separate PR for #5204 👍

nmrshll replied:
Moved it to #5888 (still in draft, I'll finish fixing this PR's feedback first)
I'll clean up what's left in this PR

muXxer commented on dev-tools/iota-private-network/bootstrap.sh:
```diff
@@ -5,12 +5,15 @@
-OVERLAY_PATH="$(pwd)/configs/validator-common.yaml"
-GENESIS_TEMPLATE="$(pwd)/configs/genesis-template.yaml"
-PRIVATE_DATA_DIR="$(pwd)/data"
+ROOT=$(git rev-parse --show-toplevel || realpath "$(dirname "$0")/../..")
```
IMO these changes in this file should not be part of this PR. We should open another PR for that.

nmrshll replied:
moved this to #5888

muXxer commented on dev-tools/iota-private-network/bootstrap.sh:
```diff
@@ -27,12 +30,13 @@ check_configs_exist() {
+    -v "$TEMP_EXPORT_DIR:/iota/configs/temp" \
     -w /iota \
-    iota-tools \
+    iotaledger/iota-tools \
```
The reasoning to not use the `iotaledger/` images was, that you have to use the "self-built" ones which can be built using the scripts in the repo. We could change everything to `iotaledger/` if we want, but then we should also change that in the build scripts, so that you either pull the latest image from docker hub, or just overwrite it with your local build. But then we should update that everywhere in the repo.

jkrvivian replied:
Now the images are tagged with `iotaledger/` prefix in the build scripts.

nmrshll replied:
moved this to #5888 and we talked about using names without `iotaledger/`, so I'll update that

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,437 @@
       "title": "Pending Futures and Tasks",
       "type": "timeseries"
+    },
+		{
```
the formatting looks a bit broken.

nmrshll replied:
fixed it

muXxer replied:
It was still broken because of a mix of tabs and spaces. I pushed a fix.

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,437 @@
+        "y": 82
+      }
+    },
+		{
```
the formatting looks a bit broken.

nmrshll replied:
fixed it

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+    pub disk_collector: IntGauge,
+    pub is_docker: bool,
+}
+impl HardwareSpecs {
```
the formatting of the file is a bit weird, sometimes there are newlines between code blocks, sometimes not.
Here there is none for example, and later in the file for some function declarations there are also none. Doesn't rustfmt handle those?

muXxer replied:
Ah no, it doesn't. Please add some newlines ;)

nmrshll replied:
added them now

nmrshll committed:
fmt

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum HardwareMetricsErr {
```
I personally would put this enum before the `register_hardware_metrics` func, but I let you decide.

nmrshll replied:
I've moved it now

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        .map_err(HardwareMetricsErr::ErrCreateMetric)?;
+        memory_collector.set(mem_total_bytes as i64);
+
+        // We're only interested in the largest disk
```
I wonder, could we actually figure out the mount point of the database folder, and search the correct disc via the `mount_point` func?

piotrm50 replied:
Or actually we could export all of them and set disk name as a label. Sometimes system has problems if a system partition is running out of storage.

nmrshll replied:
I've done both, we now have metrics for available space for each disk, as well as extra disk info to identify the DB one so we can correlate in grafana

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        let disk_total_bytes: u64 = disks
+            .iter()
+            .max_by_key(|disk| disk.total_space())
+            .map(|d| d.total_space())
```
IMO the `available_space()` is also an important metric

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        };
+        let cpu_core_count = system.physical_core_count();
+
+        let mem_total_bytes = system.total_memory();
```
IMO the `available_memory()` is also an important metric we should add.

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        metric
+    }
+    fn cpu_metric_family(&self) -> MetricFamily {
+        let mut mf = MetricFamily::new();
```
Do we even need to recreate them during every collect? Those are all fixed values, can't we simply keep the final metric family in memory and reuse that?

nmrshll replied:
I can definitely do that. 
I had it at some point, but with const labels, which didn't work at all for displaying the data in grafana.

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        metric
+    }
+    fn system_metric_family(&self) -> MetricFamily {
+        let mut mf = MetricFamily::new();
```
Do we even need to recreate them during every collect?

nmrshll committed:
revert change

nmrshll committed:
revert chnge

piotrm50 commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+        descs
+    }
+
+    fn collect(&self) -> Vec<prometheus::proto::MetricFamily> {
```
```suggestion
    fn collect(&self) -> Vec<MetricFamily> {
```

nmrshll replied:
changed already in the latest push, but good point

piotrm50 commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+                "memory_specs",
+                "Memory specs (constants: total amount, ...)",
+            )
+            .const_label("memory_total_bytes", mem_total_bytes.to_string())
```
Why are you setting `memory_total_bytes` and `memory_total_human` labels if you set the metric value to `mem_total_bytes`? Grafana can display bytes in human readable format.

nmrshll replied:
For making it easier to use at first, but I can remove this if it's entirely redundant

piotrm50 commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,280 @@
+            .unwrap_or(0);
+        let disk_collector = IntGauge::with_opts(
+            Opts::new("disk_specs", "Disk specifications (total disk space, ...)")
+                .const_label("disk_total_bytes", disk_total_bytes.to_string())
```
Same here with the labels and metric value.

piotrm50 commented on crates/iota-node/src/main.rs:
```diff
@@ -72,8 +73,9 @@ fn main() {
     let runtimes = IotaRuntimes::new(&config);
     let metrics_rt = runtimes.metrics.enter();
-    let registry_service = iota_metrics::start_prometheus_server(config.metrics_address);
+    let mut registry_service = iota_metrics::start_prometheus_server(config.metrics_address);
```
Is `mut` really necessary here? Other metrics were registered without this keyword.

nmrshll replied:
removed it 👍 
I didn't realise at first that I could register and update metrics/registries with just a `&self` ref

nmrshll committed:
re-introduce dyn usage metrics (cpu, memory, disk)

nmrshll committed:
usage for all disks

nmrshll committed:
update example dashboard

nmrshll committed:
clippy

nmrshll committed:
pr fixes

nmrshll committed:
fmt

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

nmrshll committed:
clippy

nmrshll committed:
fmt

nmrshll committed:
try fix test

nmrshll committed:
try fix tests

nmrshll committed:
try fix tests

nmrshll committed:
cleanup

nmrshll committed:
revert changes moved to #5888

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

nmrshll committed:
put error at the top

nmrshll committed:
diffferent formatting

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

muXxer committed:
fix: validator dashboard format

muXxer committed:
fix: initialize hw metrics in `start_async` to enable them in `iota-swarm`

muXxer committed:
feat: add `hw` prefix for hardware metrics

muXxer committed:
fix: small cleanups

muXxer commented:
I pushed some fixes. Now the metrics also work when run via `iota-swarm`.

Please also implement the following changes:
- all other disks than "db_disk" should also have a disk spec including their `total_bytes` in the description.
- the value of "disk spec"'s can be set to 0, same for `hw_memory_specs`
- maybe rename `db_disk_name` and `db_disk_total_bytes`, remove the `db_disk` prefix (this way you can reuse the same schema for all disc specs)

nmrshll committed:
fix PR feedback

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

nmrshll committed:
remove cpu usage; collect only disk available

nmrshll committed:
cpu specs: use cores as value, remove prefix

nmrshll committed:
separate metrics for total/available; move os info in uptime

nmrshll committed:
delete submodules

nmrshll committed:
update dashboard

nmrshll committed:
fix: uptime was missing via `iota start`

nmrshll committed:
cleanup

nmrshll committed:
docs

nmrshll committed:
format

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,485 @@
+            "uid": "${Environment}"
+          },
+          "refId": "A",
+          "expr": "uptime{host=~\"$validator\", network=\"$network\"}",
```
Looks like the exact same entry is here two times, at least I don't see the difference between the node starting in line 3850 and the one in 3707.

nmrshll replied:
Sorry about this, I must have messed up when saving (and same issue for your other comment on missing stats), this is completely different from what I thought I pushed

I've fixed it now and re-added missing panels

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,485 @@
       "type": "timeseries"
+    },
+    {
+      "title": "Hardware specs",
```
I think the following stats are missing:

- `is_docker`
- all info about the disks
- all info about the RAM

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+    pub fn new(db_path: &Path) -> Result<Self, HardwareMetricsErr> {
+        let mut system = System::new_with_specifics(
+            RefreshKind::nothing()
+                .with_cpu(CpuRefreshKind::nothing().with_cpu_usage())
```
We don't need the cpu usage kind anymore IMO.

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+        let mem_total_bytes = system.total_memory();
+        let memory_total_collector = IntGauge::with_opts(Opts::new(
+            "memory_total_bytes",
+            "Memory specs (total bytes)",
```
```suggestion
            "Memory total (bytes)",
```

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    fn collect_memory_available(&self, system: &System) -> Option<Vec<MetricFamily>> {
+        let memory_available_bytes = match i64::try_from(system.available_memory()) {
```
I don't know why, but on my system the available is always displayed as 0. 🤷

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+    fn disk_has_db(disk: &Disk, db_path: &Path) -> bool {
+        db_path.starts_with(disk.mount_point())
+    }
+    fn collect_disk_available(&self, disks: &Disks) -> Vec<MetricFamily> {
```
add a newline in front

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+                let disk_num = idx + 1;
+                Self::uint_gauge(
+                    &format!("disk_{disk_num}_available_bytes",),
+                    &format!("Disk available space in bytes, for disk {disk_num}",),
```
```suggestion
                    &format!("Disk space available (bytes), for disk {disk_num}",),
```

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+                let disk_num = idx + 1;
+                Self::uint_gauge(
+                    &format!("disk_{disk_num}_total_bytes",),
+                    &format!("Disk total space in bytes, for disk {disk_num}",),
```
```suggestion
                    &format!("Disk space total (bytes), for disk {disk_num}",),
```

muXxer commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,402 @@
+                return Vec::new();
+            }
+        };
+        system.refresh_all();
```
But why do we refresh everything?
This will also refresh CPU etc, so this might be blocking, no? Can't we just refresh the available memory and the disks?

nmrshll replied:
Yes that's a good point, we only need memory now, I updated it to only refresh this.

nmrshll committed:
Update crates/iota-metrics/src/hardware_metrics.rs

Co-authored-by: muXxer <mux3r@web.de>

nmrshll committed:
Update crates/iota-metrics/src/hardware_metrics.rs

Co-authored-by: muXxer <mux3r@web.de>

nmrshll committed:
Update crates/iota-metrics/src/hardware_metrics.rs

Co-authored-by: muXxer <mux3r@web.de>

nmrshll committed:
fix PR, update dashboard

nmrshll committed:
format

nmrshll commented:
Added these fixes (both from our call and PR feedback):
- [x] db disk: remove specs, add label  for db disk in the LABEL
- [x] total_bytes: expose as a separate metric rather than a label
- [x] drop CPU usage (too much perf impact ?)
- [x] rename hw_cpu_core_count -> value is core_count, keep string labels
- [x] remove prefix "cpu_" in labels
- [x] consistent naming with memory/disk : (one metric for _available, one metric for _total)
- [x] move os_version and is_docker to labels at top of `uptime_metric()`
- [x] uptime_metric: should also be available via iota start
- [x] updated example grafana dashboard

nmrshll committed:
Merge remote-tracking branch 'origin/develop' into 4666-add-validator-metrics-for-cpu-model-and-system-memory

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,746 @@
+          },
+          "editorMode": "code",
+          "exemplar": false,
+          "expr": "hw_disk_1_total_bytes{host=~\"$validator\",network=\"$network\",is_database_disk=\"true\"}",
```
would that even work for everyone? What if the database is on the second disk?
can we have a filter that uses the `hw_disk_XXX_total_bytes` from the disk that has `is_database_disk=\"true\"` maybe?

muXxer replied:
Does this work maybe?
`topk(1, {__name__=~"hw_disk_[0-9]+_total_bytes", host=~"$validator", network="$network", is_database_disk="true"})`

nmrshll replied:
that works better, yes
I've updated it

muXxer commented on dev-tools/grafana-local/dashboards/validator-dashboard.json:
```diff
@@ -3440,6 +3440,746 @@
+          },
+          "editorMode": "code",
+          "exemplar": false,
+          "expr": "hw_disk_1_available_bytes{host=~\"$validator\",network=\"$network\",is_database_disk=\"true\"}",
```
same here

nmrshll committed:
fix: display only database disk

muXxer committed:
Merge branch 'develop' of github.com:iotaledger/iota into 4666-add-validator-metrics-for-cpu-model-and-system-memory

muXxer committed:
chore: update Cargo.lock

alexsporn commented on crates/iota-metrics/src/hardware_metrics.rs:
```diff
@@ -0,0 +1,401 @@
+    fn collect_cpu_specs(system: &System) -> MetricFamily {
+        Self::uint_gauge(
+            "cpu_core_count",
+            "CPU core count (and labels: model,vendor,arch)",
```
`vendor_id` instead of `vendor`

muXxer committed:
fix: cpu_core_count help text
